### ISSUE TITLE
stethoscope v2

### ISSUE THREAD

**Issue by zeffii** [Sverchok/Sverchok#44]
i propose not having the draw function inside the callback class, but pass it like this 
https://github.com/nortikin/sverchok/blob/5c052118b8b1f42c7d4bc8821ad3ca855d1ddde0/nodes/number/easing.py#L147-L179

```python
        # end early
        nvBGL2.callback_disable(n_id)

        ...
   
        if self.activate:

            x, y = [int(j) for j in (self.location + Vector((self.width + 20, 0)))[:]]
            
            draw_data = {
                'tree_name': self.id_data.name[:],
                'custom_function': simple_grid_xy,
                'loc': (x, y),
                'args': (easing_func, palette)  # or whatever args
            }
            nvBGL2.callback_enable(n_id, draw_data)
```

**Comment by zeffii** [comment]
We can define the drawing code in a different file, and use the nvBGL2 (or massively stripped down version) to act only as a _facilitator_.

**Comment by zeffii** [comment]
in nvBGL2 the biggest change would be 
```python

...  import tree

def restore_opengl_defaults():
    bgl.glLineWidth(1)
    bgl.glDisable(bgl.GL_BLEND)
    bgl.glColor4f(0.0, 0.0, 0.0, 1.0)        


def draw_callback_px(n_id, data):

    space = bpy.context.space_data
  
    ng_view = space.edit_tree
    # ng_view can be None
    if not ng_view:
        return
    ng_name = space.edit_tree.name
    if not (data['tree_name'] == ng_name):
        return
    if not isinstance(ng_view, tree.SverchokReduxTree):
        return

    drawing_func = data.get('custom_function')
    x, y = data.get('loc', (20, 20))
    args = data.get('args', (None,))
    drawing_func(x, y, args)
    restore_opengl_defaults()
```

**Comment by ly29** [comment]
Basiclly all data is stored in tree structure:

A tree instance is either a leaf or a tree, if data is stored it is stored in ```tree.data``` otherwise one can access the child tree is with ```for t in tree.children```

Sample recursive function:
```python
    def count(self):
        if self.is_leaf:
            return 1
        else:
            self.obj_count = sum(t.count() for t in self.children)
            return self.obj_count
```

Pass the draw function seems like fine approach.

https://github.com/Sverchok/Sverchok/blob/master/core/data_tree.py

**Comment by ly29** [comment]
The whole structure of data tree is in need of some cleanup actually.

**Comment by zeffii** [comment]
in a few hours i'll have some time to get stuck into this - i think.

**Comment by ly29** [comment]
Give it a go.

I will add access for data tree output sockets via the socket class.

**Comment by ly29** [comment]
```
>>> data_tree=D.node_groups['NodeTree'].nodes['Linspace.005'].inputs['Start'].data_tree
>>> data_tree.level
4
```

Fun stuff
```python
list(data_tree) # give list with all children
```

**Comment by zeffii** [comment]
finding a few minutes now : ]

**Comment by zeffii** [comment]
```
 File "\svrx\nodes\output\stethoscope.py", line 57, in stop
    n_id = node_id(self)
  File "\svrx\nodes\output\stethoscope.py", line 11, in node_id
    if not self.n_id:
AttributeError: 'InnerStateful' object has no attribute 'n_id'
```
closer..

**Comment by zeffii** [comment]
or maybe it can be simpler..like https://github.com/Sverchok/Sverchok/blob/master/nodes/debug/debugprint.py

**Comment by zeffii** [comment]
maybe later.

**Comment by ly29** [comment]
It cannot really be done as a function inside the system, but should be outside of the system looking in.

**Comment by zeffii** [comment]
yeah, that's scrapped... 
```
    'tree_name': self.id_data.name[:],
AttributeError: 'InnerStateful' object has no attribute 'id_data'
```

**Comment by zeffii** [comment]
it's better i figure this out myself. :)

**Comment by ly29** [comment]
Or collect the node on init, access the socket data on stop. 
The call is just pass

**Comment by ly29** [comment]
I agree that it is better figure things out for yourself in this case I will provide some helpful hints, because I havn't made any documentation yet, and reading code that is split up into different files isn't always the easiest to track down the why and how...

In general what happens in stateful/node_func is totally separated from the blender node layout. This isn't any standard way to access the data or setting properties in the nodes for node_func, stateful has away around this.

The way to do that is to write a custom class like node script does providing the needed functionality.

The signature of ```Stateful.__init__``` is ```__init__(self, node=None)``` for *reasons* (almost). The idea being that a stateful class should be inspectable to provide the data needed for a class and function inspection using the same machinery as a ```node_func```

If there isn't a node no further call will be made to the stateful class instance.

So is this a sensible approach? That can be argued.

For doing what stethoscope want something like following is what is needed.

```python

@stateful
class SvRxStethoscope():
    bl_idname = "SvRxStethoscope"
    label = "Stethoscope"

    properties = {
        'activate': BoolP(name='activate', default=True),
        'n_id': StringP(default='callback stuff'), #  until something like n_id is implemented
    }

    def __init__(self, node=None):
        if node is not None:
            self.node = node
            self.activate = node.activate
            self.n_id = node.n_id


    @property
    def xy_offset(self):
        a = self.node.location[:]
        b = self.node.width[:]
        b[0] += 20
        return int(a[0] + b[0]), int(a[1] + b[1])

    def stop(self):
        n_id = self.n_id
        bgl_callback.callback_disable(n_id)
        data_tree = node.inputs[0].data_tree
        if self.activate:
            draw_data = make_draw_data(data_tree) # purely fictional function
            bgl_callback.callback_enable(n_id, draw_data)

    def __call__(self, data: Anytype = Required):
        pass
```

**Comment by ly29** [comment]
What should be done is almost certainly that the properties should be automatically set for a ```stateful``` node but since so few of them have been written yet I have been holding off on doing stuff while waiting for patterns to emerge rather than trying to predict them.

Then ```__init__``` would become.

Or basically not needed.

```python
    def __init__(self, node=None):
        super().__init__(node)

        
```

**Comment by ly29** [comment]
The relationship looks something like this
```
node.compile() -> callable object (node_func, node_script, stateful)
```
However since the actual node is generated from the callable is it is inspected before a node class is created.

- Node func gets their own custom class as node (but it might be shared among other node funcs via multi mode)
- All node scripts uses one class
- Stateful class have a corresponding node class.
- Stateful instances implement four methods with null defaults for two, ```start``` and ```stop```
```python
__init__(self, node=None) # 
start(self) # called when before first call to the __call__(self, arguments) 
stop(self) # called after the last call to __call__
__call__(self, arguments) # implements the callable interface common with node_func/node_script
```

The lifetime of a stateful object isn't defined/promised, right now it is very short but should/could be longer. It should be able to generate all needed data from the node anyway, since the actual node is the persistent object.

(Okay so I partly writing this out as preparation for documentation/thinking out load about the design)

**Comment by zeffii** [comment]
I will have to reread that a few times, but that the class is not a Node is understood
![image](https://cloud.githubusercontent.com/assets/619340/22877487/e40eaaa8-f1d5-11e6-8ef5-ee162c7e4857.png)

**Comment by zeffii** [comment]
now with all these data types having a `__repr__` there's a better chance of the stethoschope doing some fancy dataviz.

**Comment by ly29** [comment]
Also being able determine the border between data and structure in a clear way makes things easier as well. I am aware the this in some ways is a regression compared to the old style everything is list but I think the advantages outweigh that.

That said I am not looking forward to implementing the old list nodes.

There should also be some clever way of getting the type of data tree.

**Comment by zeffii** [comment]
hey, any new nodes with modes should use increments of 10 orso in the `id` just incase we want to add new ones in a more logical place.

We should also retrospectively correct the current `ids` to have more spaces, before it gets too late.

**Comment by ly29** [comment]
That is a good point.

**Comment by zeffii** [comment]
oh.. this needs a 
```
    def free(self):
        bgl_callback.callback_disable(node_id(self))
```
somewhere

**Comment by ly29** [comment]
Yeah, time to write a node class for this

**Comment by ly29** [comment]
NodeID is added and in stethoscope

**Comment by ly29** [comment]
The new structure of OpenGL callback really makes things simpler

**Comment by zeffii** [comment]
yeah, it'll probably work for indexviewer too,  i think it's time to copy over some 3dview bgl code :)

let me sync up.

**Comment by zeffii** [comment]
tho.. i think it would be nice to a have a mode switch for the IndexViewer.

mode bm / mode rxdata  with appropriate sockets for each.  but i'll start with bm.

**Comment by ly29** [comment]
I used the Sverchok index viewer when I made topology file in util

**Comment by ly29** [comment]
Check the stethoscope file for templates how to use a class with node_id and free

**Comment by zeffii** [comment]
yeah read the node_id new.

(i also used the sverchok viewer index to figure out wtf i was doing with the torus )

**Comment by zeffii** [comment]
i don't know how anyone uses the built in vertex index display in bpy.app.debug=True.

**Comment by nortikin** [comment]
one thought
if viewer can have adaptive sockets, so user can plug any socket no matter which, node understand what it is. Maybe except matrix.
bmesh/verts/pols/edges + matrices

**Comment by ly29** [comment]
That is the idea.

**Comment by ly29** [comment]
![stetho](https://cloud.githubusercontent.com/assets/6241382/23209534/df40be50-f8f9-11e6-97ca-cbb83b41dfbf.PNG)
First rough test.

**Comment by ly29** [comment]
![stetho2](https://cloud.githubusercontent.com/assets/6241382/23209858/4c17754a-f8fb-11e6-8189-c1ebdd13e3b9.PNG)
Somewhat disappointed in ```pprint``` here

**Comment by ly29** [comment]
Still a usable start

**Comment by ly29** [comment]
![advanced](https://cloud.githubusercontent.com/assets/6241382/23210277/2807a0d8-f8fd-11e6-9bd0-46962cdac662.PNG)
Show casing more advanced data structures

**Comment by zeffii** [comment]
the default printing for np array is pretty good, maybe we can drop pprint here

**Comment by zeffii** [comment]
if you add an index selector for stethoscope,  we can view any object,..and if it's a 100*100 array it will only print the corners.

**Comment by ly29** [comment]
I want write a more clever parser but just used ```print``` to get started. Yeah array is pretty clever

**Comment by zeffii** [comment]
like
```
c0r0, c1r0, c2r0 ...............cNr0
c0r1, c1r1, c2r1 ...............cNr1
c0r2, c1r2, c2r2 ...............cNr2
c0r3, c1r3, c2r3 ...............cNr3
....................................
c0rN, c1rN, c2rN ...............cNrN
```

**Comment by ly29** [comment]
This is actually really help already, will continue with but not right now.
